### DISCUSSION TITLE
How to access environment variable on vercel?

### DISCUSSION THREAD

**Discussion by mukeshprajapati-kombee** [vercel/community#2036] [Help] [answered]
### Summary

I have created environment variables on vercel UI but env variables are not available on production, giving me an undefined error. Does it available only for server side languages like, Node.js, Go, Python and Ruby.

### Additional information

I am accessing these variables in vue app.

**Comment by amyegan** [comment, marked as answer]
By default, environment variables are for server side use. If you want to make a non-sensitive env var available to the client, you can add a prefix appropriate to the frontend framework you are using. This standard is intended to avoid accidentally exposing a private key. 

Client-side [Vue environment variables](https://cli.vuejs.org/guide/mode-and-env.html#environment-variables) need to start with `VUE_APP_` to be embedded in the client bundle. Vercel's [System Environment Variables](https://vercel.com/docs/concepts/projects/environment-variables/system-environment-variables#system-environment-variables) are also available as [Framework Environment Variables](https://vercel.com/docs/concepts/projects/environment-variables/system-environment-variables#framework-environment-variables) with the `VUE_APP_` prefix.

**Comment by mukeshprajapati-kombee** [comment]
@amyegan Thanks for the answer. I am accessing the vercel framework environment variable but unable to access it. 

![image](https://user-images.githubusercontent.com/59353507/231982184-c63b4bd4-2dff-4ebf-ac52-f8860f06899d.png)
![image](https://user-images.githubusercontent.com/59353507/231982371-d257858f-0b8c-4247-be15-f8571625dfc0.png)

**Reply by amyegan** [reply]
If your project uses Vite, the [env variables](https://vitejs.dev/guide/env-and-mode.html#env-variables-and-modes) may be on a special `import.meta.env` object. In that case, you may also need to use the `VITE_` prefix.

**Comment by mukeshprajapati-kombee** [comment]
@amyegan I am using vue/quasar app. I am not using VITE

**Reply by amyegan** [reply]
Since you're using Quasar, please refer to their documentation on [handling `process.env` with Vite](https://quasar.dev/quasar-cli-vite/handling-process-env) or [handling `process.env` with Webpack](https://quasar.dev/quasar-cli-webpack/handling-process-env/), depending on which build tool is being used. If you haven't already, you should also reach out in the [Quasar forum](https://forum.quasar.dev/). The folks over there are much more familiar with the framework and can give more specific recommendations

**Reply by mukeshprajapati-kombee** [reply]
@amyegan Thanks, I will ask there,

**Comment by mukeshprajapati-kombee** [comment]
@amyegan I have one more question, how I can enable maintenance mode in my deployed quasar app using vercel?

**Reply by amyegan** [reply]
You can use Edge Middleware to enable maintenance mode. You can find more info in the [Redirects: Edge Middleware](https://vercel.com/docs/concepts/edge-network/redirects#edge-middleware) documentation. There's also a [maintenance page template](https://vercel.com/templates/next.js/maintenance-page) and [example repo](https://github.com/vercel/examples/tree/main/edge-middleware/maintenance-page) available so you can see the code

**Comment by mukeshprajapati-kombee** [comment]
@amyegan Is it required for Next/js or any other SSR framework? I can see next.js is used in the example. 

I am getting the following error when accessing the edge config key-value pair in my quasar/vue SPA app:

![image](https://user-images.githubusercontent.com/59353507/233288458-3951a9ee-a847-42de-9f5b-9ad0c0d803f1.png)

**Reply by amyegan** [reply]
Next.js is not a requirement for middleware and edge config. After creating the Edge Config store and key-value pair, don't forget to [pull the latest environment variables](https://vercel.com/docs/concepts/edge-network/edge-config/get-started#pull-the-latest-environment-variables) into your local environment.

**Reply by mukeshprajapati-kombee** [reply]
@amyegan Yeah, I have pulled envs into my local.

**Reply by mukeshprajapati-kombee** [reply]
@amyegan Connection string is also appending with the get request of edge config.

**Reply by amyegan** [reply]
If you can share a minimal reproduction, I would be happy to take a closer look

**Reply by mukeshprajapati-kombee** [reply]
Sure, this is the GitHub repo: https://github.com/mukeshghodela-easternts/quasar-pinia-boilerplate
Latest code is pushed in the `develop` branch.
Live URL for testing : https://quasar-pinia-boilerplate-git-develop-mukeshghodela-easternts.vercel.app/#/
I have added the `EDGE_CONFIG` environment variable in the `quasar.config.js` file.

![image](https://user-images.githubusercontent.com/59353507/233896226-4b02a9df-9736-4751-8762-2290a01fd8b2.png)

You can find the middleware file under:  `src\middlewares\middleware.ts`

![image](https://user-images.githubusercontent.com/59353507/233896346-8b2e8e7f-6270-4e27-9953-5fc49291b3ae.png)

I am calling middleware code in the route file: `src\router\index.ts`

![image](https://user-images.githubusercontent.com/59353507/233896516-b6c958aa-106e-402a-afb3-0a9116bf4cea.png)

EDGE Config API is also calling: 

![image](https://user-images.githubusercontent.com/59353507/233896796-49ce2a61-2725-40c1-882d-df0b9057ef40.png)

**Reply by amyegan** [reply]
I believe [`middleware.js` needs to be in the root or `/src` folder](https://vercel.com/docs/concepts/functions/edge-middleware#:~:text=You%20can%20use%20Edge%20Middleware%20with,in%20the%20root%20of%20your%20project). If you look at the deployment on your Vercel project dashboard and go to the "Source" tab, you will see which files were provided for the build. I'm getting quasar errors when I try to build the sample project, so you may have a different result, but I'm not seeing the middleware.js file come through in the build files. That makes me think a configuration change may be needed to get quasar to include the file.

**Reply by mukeshprajapati-kombee** [reply]
@amyegan middleware is there in the source 
![image](https://user-images.githubusercontent.com/59353507/234196393-6e73b24b-97c6-414b-b739-4418eda836bf.png)

**Reply by amyegan** [reply]
I'm sorry I didn't notice this before. Your middleware.js is not returning a redirect or rewrite to the maintenance page. For Vercel Middleware to send users to a different page, the function would need to return something like in this [middleware code example](https://vercel.com/docs/concepts/functions/edge-middleware/quickstart#add-middleware-code). In that case, the middleware should handle the redirect automatically and you wouldn't need to use `Router.beforeEach`.

If it makes more sense to have Quasar handle the routing, then you could use the Edge Config without Edge Middleware to pass along the config value. The docs on [Reading Edge Configs with the SDK](https://vercel.com/docs/concepts/edge-network/edge-config/edge-config-sdk#) explain in more detail. There's also a brief [troubleshooting](https://vercel.com/docs/concepts/edge-network/edge-config/edge-config-limits#troubleshooting) guide.

**Reply by mukeshprajapati-kombee** [reply]
@amyegan Quasar will not handle the routing of edge middleware. Will edge middle ware invoke automatically on any route?

**Reply by amyegan** [reply]
Edge Middleware runs on every request by default, but a [`config` object](https://vercel.com/docs/concepts/functions/edge-middleware/middleware-api#config-object) can be used to match one or more routes and limit where Middleware should run. You can find more info in the [Edge Middleware API docs](https://vercel.com/docs/concepts/functions/edge-middleware/middleware-api#match-paths-based-on-custom-matcher-config)

**Reply by mukeshprajapati-kombee** [reply]
@amyegan Thanks I will follow the docs

**Comment by gederer** [comment]
I'm trying to use Vercel Postgres in a Qwik app. I can see all of the Postgres-related environment variables in the Vercel admin UI under Project Settings/Environment Variables. But, when I have:

```typescript
  const client = await db.connect();
  const roles = await client.sql`SELECT * FROM "Role"`;

  return roles.rows;
```

I get an error telling me that `POSTGRES_URL` is not set. Also, when I `console.log(process.env)`, the Postgres-related variables are absent.

The app works as expected when running on my local machine using the `.env` file that I grabbed with `vercel env pull .env.development.local`. Any suggestions?

Note: Qwik uses Vite, which might (or might not) be a useful clue. I know it's kind of new, but a lot of folks are checking it out just now. Cheers.
